ts41/get.d.ts

```diff
@@ -0,0 +1,123 @@
+import {Split} from './utilities';
+import {StringDigit} from '../source/utilities';
+
+/**
+Like @see Get but receives an array of strings as a path parameter.
+*/
+type GetWithPath<ObjectType, Keys extends readonly string[]> =
+	Keys extends []
+	? ObjectType
+	: Keys extends [infer Head, ...infer Tail]
+	? GetWithPath<PropertyOf<ObjectType, Extract<Head, string>>, Extract<Tail, string[]>>
+	: never;
+
+/**
+Splits a dot-prop style path into a tuple comprised of the properties in the path. Handles square-bracket notation.
+
+@example
+```
+ToPath<'foo.bar.baz'>
+//=> ['foo', 'bar', 'baz']
+
+ToPath<'foo[0].bar.baz'>
+//=> ['foo', '0', 'bar', 'baz']
+```
+*/
+type ToPath<S extends string> = Split<FixPathSquareBrackets<S>, '.'>;
+
+/**
+Replaces square-bracketed dot notation with dots, for example, `foo[0].bar` -> `foo.0.bar`.
+*/
+type FixPathSquareBrackets<Path extends string> =
+	Path extends `${infer Head}[${infer Middle}]${infer Tail}`
+	? `${Head}.${Middle}${FixPathSquareBrackets<Tail>}`
+	: Path;
+
+/**
+Returns true if `S` is made up out of `C` repeated 0 or more times.
+
+@example
+```
+ConsistsOnlyOf<'aaa', 'a'> //=> true
+ConsistsOnlyOf<'aBa', 'a'> //=> false
+ConsistsOnlyOf<'', 'a'> //=> true
+```
+*/
+type ConsistsOnlyOf<LongString extends string, Substring extends string> =
+	LongString extends ''
+	? true
+	: LongString extends `${Substring}${infer Tail}`
+	? ConsistsOnlyOf<Tail, Substring>
+  : false;
+
+/**
+Convert a type which may have number keys to one with string keys, making it possible to index using strings retrieved from template types.
+
+@example
+```
+type WithNumbers = {foo: string; 0: boolean};
+type WithStrings = WithStringKeys<WithNumbers>;
+
+type WithNumbersKeys = keyof WithNumbers;
+//=> 'foo' | 0
+type WithStringsKeys = keyof WithStrings;
+//=> 'foo' | '0'
+```
+*/
+type WithStringKeys<ObjectType extends Record<string | number, any>> = {
+	[Key in `${Extract<keyof ObjectType, string | number>}`]: ObjectType[Key]
+};
+
+/**
+Get a property of an object or array. Works when indexing arrays using number-literal-strings, for example, `PropertyOf<number[], '0'> = number`, and when indexing objects with number keys.
+
+Returns `unknown` if `Key` is not a property of `ObjectType`.
+*/
+type PropertyOf<ObjectType, Key extends string> =
+	ObjectType extends null | undefined
+	? undefined
```

Comment:
I wrote this as a model for the `get` functions from libraries like lodash, dot-prop and jest's `expect({ foo: { bar: 1 } }).toHaveProperty('foo.bar', 1)`. For that behaviour to be properly mapped to the type system, nullish values need to propagate:

```js
declare const myObj: {
  foo?: {
    bar: { baz: number }
  }
}

get(myObj, 'foo.bar.baz')
```

The return type for the expression above should be `number | undefined`, but if we changed this to `never`, the return type would be `number` (because `number | never` => `number`). Which would be an accurate model for a library which _throws_ if the path can't be fully resolved, but I'm not aware of any that work that way.